overlord/snapstate/component.go

```diff
@@ -0,0 +1,227 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapstate
+
+import (
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/overlord/snapstate/backend"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+// InstallComponentPath returns a set of tasks for installing a snap component
+// from a file path.
+//
+// Note that the state must be locked by the caller. The provided SideInfo can
+// contain just a name which results in local sideloading of the component, or
+// full metadata in which case the component will appear as installed from the
+// store.
+func InstallComponentPath(st *state.State, csi *snap.ComponentSideInfo, si *snap.SideInfo,
+	path, snapInstance, channel string, flags Flags) (*state.TaskSet, error) {
```

Comment:
this should probably take a snap.Info instead of snapInstance and si ?

Reply:
I don't think we can change a channel just by installing a component, unless we are doing that for the snap overall, so this probably shouldn't take a channel at all for now